device.rs: increment metric instead of logging error

iulianbarbu opened the pull request:
We've introduced an error log for the case when there is no available buffer for RX data. This can happen for heavy network traffic usage, and it is not an actual error, rather a limitation for which we must increment a metric.

Signed-off-by: iulianbarbu <iul@amazon.com>

## Reason for This PR

Fixes #1677 

## Description of Changes

Removed an unnecessary error log.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.

acatangiu commented on src/devices/src/virtio/net/device.rs:
```diff
@@ -543,7 +543,6 @@ impl Net {
         METRICS.net.rx_tap_event_count.inc();
         if self.queues[RX_INDEX].is_empty(&self.mem) {
-            error!("The RX queue is empty, there is no available buffer.");
             METRICS.net.event_fails.inc();
```
I suggest we use a specific metric for this case. This metric is also used for error-paths.

acatangiu commented on src/devices/src/virtio/net/device.rs:
```diff
@@ -543,7 +543,6 @@ impl Net {
     pub fn process_tap_rx_event(&mut self) {
         METRICS.net.rx_tap_event_count.inc();
         if self.queues[RX_INDEX].is_empty(&self.mem) {
-            error!("The RX queue is empty, there is no available buffer.");
```
Same problem in line 575 and I expect others. Please have the fix address all instances of this behavior.

andreeaflorescu commented on src/devices/src/virtio/net/device.rs:
```diff
@@ -652,7 +639,6 @@ impl VirtioDevice for Net {
     fn read_config(&self, offset: u64, mut data: &mut [u8]) {
         let config_len = self.config_space.len() as u64;
         if offset >= config_len {
-            error!("Failed to read config space");
```
Reading and writing the config space should not be on the critical path. We should keep this log and the next one.

acatangiu commented on src/devices/src/virtio/block/device.rs:
```diff
@@ -144,7 +144,6 @@ impl Block {
     pub(crate) fn process_queue_event(&mut self) {
         METRICS.block.queue_event_count.inc();
         if let Err(e) = self.queue_evts[0].read() {
-            error!("Failed to get queue event: {:?}", e);
```
This is an actual error that shouldn't happen. Keep it.

acatangiu commented on src/devices/src/virtio/block/device.rs:
```diff
@@ -204,7 +203,6 @@ impl Block {
                             VIRTIO_BLK_S_OK
                         }
                         Err(e) => {
-                            error!("Failed to execute request: {:?}", e);
```
Also valid error that shouldn't happen in a valid scenario. Keep the log.

acatangiu commented on src/devices/src/virtio/block/device.rs:
```diff
@@ -215,7 +213,6 @@ impl Block {
                     self.mem.write_obj(status, request.status_addr).unwrap();
                 }
                 Err(e) => {
-                    error!("Failed to parse available descriptor chain: {:?}", e);
```
Keep it

acatangiu commented on src/devices/src/virtio/block/device.rs:
```diff
@@ -232,7 +229,6 @@ impl Block {
             .fetch_or(VIRTIO_MMIO_INT_VRING as usize, Ordering::SeqCst);
 
         self.interrupt_evt.write(1).map_err(|e| {
-            error!("Failed to signal used queue: {:?}", e);
```
Keep it

acatangiu commented on src/devices/src/virtio/block/device.rs:
```diff
@@ -290,7 +286,6 @@ impl VirtioDevice for Block {
     fn read_config(&self, offset: u64, mut data: &mut [u8]) {
         let config_len = self.config_space.len() as u64;
         if offset >= config_len {
-            error!("Failed to read config space");
```
keep it

iulianbarbu commented:
@acatangiu , @andreeaflorescu , I've addressed your feedback! Per my current analysis, I've added metrics for no available buffers case, for the block & net devices (separate metrics for rx and tx queues).

acatangiu commented on src/devices/src/virtio/net/device.rs:
```diff
@@ -543,8 +542,7 @@ impl Net {
         if self.queues[RX_INDEX].is_empty(&self.mem) {
-            error!("The RX queue is empty, there is no available buffer.");
-            METRICS.net.event_fails.inc();
+            METRICS.net.no_rx_avail_buffer.inc();
```
You need to increment this metric in _Line 225_ too.

```rust
    fn rx_single_frame(&mut self) -> bool {
        let rx_queue = &mut self.queues[RX_INDEX];
        let mut next_desc = rx_queue.pop(&self.mem);
        if next_desc.is_none() {
            // No buffers available <---------------------- increment metric
            return false;
        }
        ...
    }
```

iulianbarbu replied:
This is true. I'll increment the no_rx_avail_buffer metric here, too.

iulianbarbu commented:
@acatangiu , @andreeaflorescu , please take a look again. I had to update the tests/integration_tests/build/test_coverage.py.
